mlflow/pyfunc/__init__.py

```diff
@@ -1648,12 +1662,160 @@ def _convert_struct_values(
+        Databricks runtime or Databricks Connect client.
+        The prebuilt env archive file can't be used across different Databricks runtime
+        versions or different platform machines.
+        When using the prebuilt env in `mlflow.pyfunc.spark_udf`, MLflow will verify
+        whether the spark UDF sandbox environment matches the prebuilt env requirements.
+
+    .. code-block:: python
+        :caption: Example
+
+        from mlflow.pyfunc import prebuild_model_env
+
+        # Create a python environment archive file at the path `prebuilt_env_path`
+        prebuilt_env_path = prebuild_model_env(
+            f"runs:/{run_id}/model", "/path/to/save_directory"
+        )
+
+    Args:
+        model_uri: URI to the model that is used to build the python environment.
+        save_path: The directory path that is used to save the prebuilt model environment
+            archive file path.
+    """
+    from mlflow.utils._spark_utils import _get_active_spark_session
+
+    if not is_in_databricks_runtime():
+        raise RuntimeError("'prebuild_model_env' only support running in Databricks runtime.")
+
+    if not os.path.exists(save_path):
+        raise RuntimeError(f"The saving path '{save_path}' does not exist.")
+    if not os.path.isdir(save_path):
+        raise RuntimeError(f"The saving path '{save_path}' must be a directory.")
+
+    local_model_path = _download_artifact_from_uri(
+        artifact_uri=model_uri, output_path=_create_model_downloading_tmp_dir(should_use_nfs=False)
+    )
+    archive_name = _gen_prebuilt_env_archive_name(_get_active_spark_session(), local_model_path)
+    dest_path = os.path.join(save_path, archive_name + ".tar.gz")
+    if os.path.exists(dest_path):
+        raise RuntimeError(
+            "You have pre-built the model python environment and saved "
+            f"it in the '{save_path}' directory as the archive file "
+            f"{archive_name}.tar.gz. To rebuild it, please remove "
+            "the existing one first."
+        )
+
+    # Archive the environment directory as a `tar.gz` format archive file,
+    # and then move the archive file to the destination directory.
+    # Note:
+    # - all symlink files in the input directory are kept as it is in the
+    #  archive file.
+    # - the destination directory could be UC-volume fuse mounted directory
+    #  which only supports limited filesystem operations, so to ensure it works,
+    #  we generate the archive file under /tmp and then move it into the
+    #  destination directory.
+    tmp_archive_path = None
+    try:
+        tmp_archive_path = _prebuild_env_internal(
+            local_model_path, archive_name, _PREBUILD_ENV_ROOT_LOCATION
+        )
+        shutil.move(tmp_archive_path, save_path)
+        return dest_path
+    finally:
+        shutil.rmtree(local_model_path, ignore_errors=True)
+        if tmp_archive_path and os.path.exists(tmp_archive_path):
+            os.remove(tmp_archive_path)
 
 
 def spark_udf(
```

Comment:
`spark_udf` looks very complex now. Can we refactor in a follow-up?

Reply:
yes, but need to think how to refactor.